### PR TITLE
BUG: SeriesGroupBy.value_counts - index name missing in categorical columns

### LINKED ISSUE #44324: BUG: SeriesGroupBy.value_counts - index name missing when applied on categorical column
### 

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [ ] I have confirmed this bug exists on the master branch of pandas.

### Reproducible Example

```python
import pandas as pd
df = pd.DataFrame({'gender': ['female', 'male', 'female', 'male', 'female', 'male'],
                  'education': ['low', 'medium', 'high', 'low', 'high', 'low'],
                  'country': ['US', 'FR', 'US', 'FR', 'FR', 'US']})
df.dtypes  
gender       object
education    object
country      object
dtype: object

df.groupby('country')['gender'].value_counts()

country  gender
FR       male      2
         female    1
US       female    2
         male      1
Name: gender, dtype: int64

df.groupby('country')['gender'].value_counts().index.names

FrozenList(['country', 'gender'])
```

Now, if using a categorical type on the genre column, the column name is not saved in the index:

```python
df['gender'] = df['gender'].astype('category')
df.dtypes

gender       category
education      object
country        object
dtype: object

df.groupby('country')['gender'].value_counts()

country
FR       male      2
         female    1
US       female    2
         male      1
Name: gender, dtype: int64

df.groupby('country')['gender'].value_counts().index.names

FrozenList(['country', None])
```

### Issue Description

The column name (here **gender**) is dropped when applying `value_counts()` on a column of type `category`.

It might be link with other problems in handling categorical types: #44001 

This was working as of pandas v 1.2.3, and I could reproduce the bug starting with v 1.3.0, still present in v 1.3.4

### Expected Behavior

Column name should be added in the index for uniformity with other types, as in the first example above with object type column.

### Installed Versions

<details>
pd.show_versions() 
INSTALLED VERSIONS
------------------
commit           : 945c9ed766a61c7d2c0a7cbb251b6edebf9cb7d5
python           : 3.9.6.final.0
python-bits      : 64
OS               : Darwin
OS-release       : 21.1.0
Version          : Darwin Kernel Version 21.1.0: Wed Oct 13 17:33:23 PDT 2021; root:xnu-8019.41.5~1/RELEASE_X86_64
machine          : x86_64
processor        : i386
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : None.UTF-8

pandas           : 1.3.4
numpy            : 1.21.4
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 21.3.1
setuptools       : 57.0.0
Cython           : None
pytest           : 6.2.4
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : 4.6.3
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.1
IPython          : 7.26.0
pandas_datareader: None
bs4              : 4.9.3
bottleneck       : None
fsspec           : None
fastparquet      : None
gcsfs            : None
matplotlib       : 3.4.3
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyxlsb           : None
s3fs             : None
scipy            : 1.7.1
sqlalchemy       : None
tables           : None
tabulate         : 0.8.9
xarray           : 0.19.0
xlrd             : None
xlwt             : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by NumberPiOso**
- [X] closes  #44324
- [X] tests added / passed
- [X] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [X] whatsnew entry

### CODE REVIEW

File: pandas/core/groupby/generic.py [hunk 1]
```diff
@@ -602,14 +602,18 @@ def value_counts(
         ids, _, _ = self.grouper.group_info
         val = self.obj._values
 
+        names = self.grouper.names + [self.obj.name]
+
         def apply_series_value_counts():
-            return self.apply(
```
Comment (mroeschke) [inline pandas/core/groupby/generic.py:606]: Could you take this opportunity to refactor and inline this logic like

```
if (bins is not None and np.iterable(bins) or is_categorical_dtype(val.dtype):
    # comments
    s = self.apply(...)
    s.index.names = ...
```

Also will s always have a MultiIndex here?

Comment (NumberPiOso) [reply pandas/core/groupby/generic.py:606]: If I run the test code with the following change 

```python
    if is_categorical_dtype(val.dtype) or (bins and np.iterable(bins)):
        # scalar bins cannot be done at top level
        # in a backward compatible way
        # GH38672
        s = self.apply(
            Series.value_counts,
            normalize=normalize,
            sort=sort,
            ascending=ascending,
            bins=bins,
        )
        print(s.index)
        s.index.names = names
```

I get these results

```python
df = pd.DataFrame(
    {
        "gender": ["female"],
        "country": ["US"],
    }
)
df["gender"] = df["gender"].astype("category")
result2 = df.groupby("country")["gender"].value_counts()
MultiIndex([('US', 'female')],
           names=['country', None])
```
And I think the fact that we are inside a group by should always ensure that we are always using MultiIndex.

Running the following code calls a different method. Specifically, `pandas.core.algorithms.value_counts`.

File: pandas/core/groupby/generic.py [hunk 2]
```diff
@@ -602,23 +602,21 @@ def value_counts(
         ids, _, _ = self.grouper.group_info
         val = self.obj._values
 
-        def apply_series_value_counts():
-            return self.apply(
+        names = self.grouper.names + [self.obj.name]
+
+        if is_categorical_dtype(val.dtype) or (bins and np.iterable(bins)):
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:607]: Should this be `... and not np.iterable(bins)`?

File: pandas/core/groupby/generic.py [hunk 3]
```diff
@@ -683,7 +681,6 @@ def apply_series_value_counts():
         levels = [ping.group_index for ping in self.grouper.groupings] + [
             lev  # type: ignore[list-item]
         ]
-        names = self.grouper.names + [self.obj.name]
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:686]: Was this line just never necessary (i.e. independent of your change above)?

File: pandas/core/groupby/generic.py [hunk 4]
```diff
@@ -602,23 +602,21 @@ def value_counts(
...
         val = self.obj._values
 
-        def apply_series_value_counts():
-            return self.apply(
+        names = self.grouper.names + [self.obj.name]
+
+        if is_categorical_dtype(val.dtype) or (bins and not np.iterable(bins)):
+            # scalar bins cannot be done at top level
+            # in a backward compatible way
+            # GH38672
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:610]: can you clarify that GH#38672 refers to the categorical case

File: pandas/core/groupby/generic.py [hunk 5]
```diff
@@ -602,23 +602,21 @@ def value_counts(
...
-        if bins is not None:
-            if not np.iterable(bins):
-                # scalar bins cannot be done at top level
-                # in a backward compatible way
-                return apply_series_value_counts()
-        elif is_categorical_dtype(val.dtype):
-            # GH38672
-            return apply_series_value_counts()
+            s.index.names = names
+            return s
```
Comment (jbrockmendel) [inline pandas/core/groupby/generic.py:619]: nitpick: s -> ser

File: pandas/core/groupby/generic.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mroeschke) [reply pandas/core/groupby/generic.py]: This looks okay but looks like tests are still failing

[the remark this answers is not in the document]
Comment (NumberPiOso) [reply pandas/core/groupby/generic.py]: Indeed, apparently I introduced the error there
```python
>       if is_categorical_dtype(val.dtype) or (bins and not np.iterable(bins)):
>      ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```

**Review by mroeschke — APPROVED** [review]

**Review by rhshadrach — APPROVED** [review]
lgtm

**Comment by mroeschke** [comment]
Thanks @NumberPiOso

**Pull request merged**